### PR TITLE
[ISSUE 11796] throw NPE when readEntry

### LINKED ISSUE #11796: throw NPE when readEntry
**function call chain：**

**ManagedCursorImpl**

```java
@Override
 public void asyncReadEntries(int numberOfEntriesToRead, long maxSizeBytes, ReadEntriesCallback callback,
                              Object ctx, PositionImpl maxPosition) {
     checkArgument(numberOfEntriesToRead > 0);
     if (isClosed()) {
         callback.readEntriesFailed(new ManagedLedgerException
                 .CursorAlreadyClosedException("Cursor was already closed"), ctx);
         return;
     }

     int numOfEntriesToRead = applyMaxSizeCap(numberOfEntriesToRead, maxSizeBytes);
    
     PENDING_READ_OPS_UPDATER.incrementAndGet(this);
     OpReadEntry op = OpReadEntry.create(this, readPosition, numOfEntriesToRead, callback, ctx, maxPosition);
     ledger.asyncReadEntries(op);
 }
```

**OpReadEntry**

```java
 public static OpReadEntry create(ManagedCursorImpl cursor, PositionImpl readPositionRef, int count,
            ReadEntriesCallback callback, Object ctx, PositionImpl maxPosition) {

     OpReadEntry op = RECYCLER.get();
     op.readPosition = cursor.ledger.startReadOperationOnLedger(readPositionRef, op);
     op.cursor = cursor;
     op.count = count;
     op.callback = callback;
     op.entries = Lists.newArrayList();
     if (maxPosition == null) {
         maxPosition = PositionImpl.latest;
     }
     op.maxPosition = maxPosition;
     op.ctx = ctx;
     op.nextReadPosition = PositionImpl.get(op.readPosition);
     return op;
 }
```

**ManagedLedgerImpl**

````java
PositionImpl startReadOperationOnLedger(PositionImpl position, OpReadEntry opReadEntry) {
     Long ledgerId = ledgers.ceilingKey(position.getLedgerId());
     if (null == ledgerId) {
         opReadEntry.readEntriesFailed(new ManagedLedgerException.NoMoreEntriesToReadException("The ceilingKey(K key) method is used to return the " +
                 "least key greater than or equal to the given key, or null if there is no such key"), null);
     }

```
 if (ledgerId != position.getLedgerId()) {
     // The ledger pointed by this position does not exist anymore. It was deleted because it was empty. We need
     // to skip on the next available ledger
     position = new PositionImpl(ledgerId, 0);
 }

 return position;
```

 }
````

**OpReadEntry**

```java
@Override
public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
	cursor.readOperationCompleted();
	// ...
}
```

**description**
if read ledger not in ledgers, there is a exception. 
but OpReadEntry instance is not be initialized, variable cursor is null.
function readEntriesFailed will throw NPE, and OpReadEntry is not recycle.

### PR DESCRIPTION
**Pull request by casuallc**
Fixes #11796 

[ISSUE 11796] throw NPE when readEntry
  
No need doc.

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -762,6 +764,9 @@ public void asyncReadEntriesOrWait(int maxEntries, long maxSizeBytes, ReadEntrie
         } else {
             OpReadEntry op = OpReadEntry.create(this, readPosition, numberOfEntriesToRead, callback,
                     ctx, maxPosition);
+            if (op.readPosition == null) {
+                return;
```
Comment (gaoran10) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:768]: Maybe we need to call the method `callback.readEntriesFailed` to make sure the caller gets a result.

Comment (casuallc) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:768]: Neet not call the method here, if `OpReadEntry` init failed, must ensure `OpReadEntry->readEntriesFailed` will be called, `callback.readEntriesFailed` will be called then.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -622,9 +622,11 @@ public void asyncReadEntries(int numberOfEntriesToRead, long maxSizeBytes, ReadE
 
         int numOfEntriesToRead = applyMaxSizeCap(numberOfEntriesToRead, maxSizeBytes);
 
-        PENDING_READ_OPS_UPDATER.incrementAndGet(this);
         OpReadEntry op = OpReadEntry.create(this, readPosition, numOfEntriesToRead, callback, ctx, maxPosition);
-        ledger.asyncReadEntries(op);
+        if (op.readPosition != null) {
```
Comment (315157973) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:626]: op is not recycle if op.readPosition == null

By the way, could you add some unit test?

Comment (casuallc) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:626]: op recycled when the method `OpReadEntry->readEntriesFailed` called.

At present, op init failed only if `op.readPosition = cursor.ledger.startReadOperationOnLedger(readPositionRef, op);` called.

I do not know how to test this case ...

**Comment by lhotari** [comment]
good work @casuallc . The changes in this PR seem to be necessary for handling the failure case that is highlighted. 

btw. There seems to be a similar change #11292 in progress, but that is taking a different approach for solving a similar issue. I'd prefer the approach used in this PR over the changes in 11292 since the solution in this PR is very simple.

**Review by lhotari — APPROVED** [review]

**Comment by Jason918** [comment]
> good work @casuallc . The changes in this PR seem to be necessary for handling the failure case that is highlighted.
> 
> btw. There seems to be a similar change #11292 in progress, but that is taking a different approach for solving a similar issue. I'd prefer the approach used in this PR over the changes in 11292 since the solution in this PR is very simple.

@lhotari 
IMHO, there is race condition in this solution. 
If `op.readPosition = cursor.ledger.startReadOperationOnLedger(readPositionRef, op);` is failed in `OpReadEntry.create`, this OpReadEntry is recycled in `readEntriesFailed`. 
So if we check `op.readPosition` after `OpReadEntry.create`, this `OpReadEntry` object could be reused in other threads, and maybe the `readPosition` is properly initialized.

That's why I try to init `readPosition` after each `OpReadEntry.create` method in #11292 , causing the complexity.

**Comment by casuallc** [comment]
> > good work @casuallc . The changes in this PR seem to be necessary for handling the failure case that is highlighted.
> > btw. There seems to be a similar change #11292 in progress, but that is taking a different approach for solving a similar issue. I'd prefer the approach used in this PR over the changes in 11292 since the solution in this PR is very simple.
> 
> @lhotari
> IMHO, there is race condition in this solution.
> If `op.readPosition = cursor.ledger.startReadOperationOnLedger(readPositionRef, op);` is failed in `OpReadEntry.create`, this OpReadEntry is recycled in `readEntriesFailed`.
> So if we check `op.readPosition` after `OpReadEntry.create`, this `OpReadEntry` object could be reused in other threads, and maybe the `readPosition` is properly initialized.
> 
> That's why I try to init `readPosition` after each `OpReadEntry.create` method in #11292 , causing the complexity.

@Jason918 Yes, you are right. I should check op but not op.readPosition.
I changed the code, PTAL.

**Comment by Shoothzj** [comment]
@lhotari code `managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java` the callback paas the ctx as null.
But in class `PersistentDispatcherSingleActiveConsumer` `readEntryFailed` method acquire the ctx(Consumer) not be null.
So the code should be

`            opReadEntry.readEntriesFailed(new ManagedLedgerException.NoMoreEntriesToReadException("The ceilingKey(K key) method is used to return the " +
                    "least key greater than or equal to the given key, or null if there is no such key"), ctx);`

Am i right ?

**Comment by lhotari** [comment]
@nicoloboschi @eolivelli The changes in this PR would be useful for fixing a common NPE issue #11796.

**Comment by casuallc** [comment]
> @casuallc It looks like your branch for this PR is gone and this PR cannot be re-opened. I closed it with the intention to trigger a rebuild. Rebuilding on CI was disabled and now I can see that the reason was that the original PR branch has been deleted.
> 
> @casuallc Would you mind creating a new PR with the same changes?

#12396

**Pull request closed without merging**